api/filters/platform_termination_filter.go

```diff
@@ -62,11 +65,68 @@ func (f *platformTerminationFilter) Run(req *web.Request, next web.Handler) (*we
 				StatusCode:  http.StatusUnprocessableEntity,
 			}
 		}
+		sharingReferences, err := findReferencesOfSharedInstancesInOtherPlatforms(ctx, platform, f.repository)
+		if err != nil {
+			return nil, err
+		}
+		if len(sharingReferences) > 0 {
+			return nil, &util.HTTPError{
+				ErrorType:   "UnprocessableEntity",
+				Description: "Platform cannot be deleted because other platform(s) has reference instance(s) to the shared instances in given platform. Details: " + formatSharingReferences(sharingReferences),
+				StatusCode:  http.StatusUnprocessableEntity,
+			}
+		}
 	}
-
 	return next.Handle(req)
 }
 
+func formatSharingReferences(references []*SharingReferences) string {
+	var msg []string
+	for i := 0; i < len(references); i++ {
+		msg = append(msg, "shared instance "+references[i].sharedInstanceID+" referenced by instance(s) "+strings.Join(references[i].referenceIDs, ", "))
+	}
+	return strings.Join(msg, ", ")
+}
+
+type SharingReferences struct {
+	sharedInstanceID string
+	referenceIDs     []string
+}
+
+func findReferencesOfSharedInstancesInOtherPlatforms(ctx context.Context, platform *types.Platform, repository storage.Repository) ([]*SharingReferences, error) {
+	sharedInstances, err := repository.ListNoLabels(ctx, types.ServiceInstanceType,
+		query.ByField(query.EqualsOperator, "platform_id", platform.ID),
+		query.ByField(query.EqualsOperator, "shared", "true"),
+	)
+	if err != nil {
+		return nil, util.HandleStorageError(err, types.ServiceInstanceType.String())
+	}
+	var referencesInOtherPlatforms []*SharingReferences
+	for i := 0; i < sharedInstances.Len(); i++ {
+		sharedInstance := sharedInstances.ItemAt(i).(*types.ServiceInstance)
+		references, err := repository.ListNoLabels(
+			ctx,
+			types.ServiceInstanceType,
+			query.ByField(query.EqualsOperator, instance_sharing.ReferencedInstanceIDKey, sharedInstance.GetID()),
+			query.ByField(query.NotEqualsOperator, "platform_id", platform.GetID()),
+		)
+		if err != nil {
+			return nil, util.HandleStorageError(err, types.ServiceInstanceType.String())
+		}
+		if references.Len() > 0 {
+			var referenceIDs []string
+			for ri := 0; ri < references.Len(); ri++ {
+				referenceIDs = append(referenceIDs, references.ItemAt(ri).GetID())
+			}
+			referencesInOtherPlatforms = append(referencesInOtherPlatforms, &SharingReferences{
+				sharedInstanceID: sharedInstance.GetID(),
+				referenceIDs:     referenceIDs,
+			})
+		}
+	}
```

Comment:
Why do we need a query per instance? The same can be done with a single DB query using InOperator, e.g. ```query.ByField(query.InOperator, instance_sharing.ReferencedInstanceIDKey, types.ObjectListIDsToStringArray(sharedInstances)...)```

Reply:
update to have a single db query